### PR TITLE
sdk(react-native-sdk): rnsdk screenshare android fix

### PR DESCRIPTION
**Pull request by Calinteodor**
Check individual commits.
https://github.com/jitsi/jitsi-meet/assets/26413496/f49e09e8-68ac-4855-b47d-9caad531e6d0

### CODE REVIEW

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        channel = new NotificationChannel("JitsiOngoingConferenceChannel", "ongoing_notification_action_unmute", NotificationManager.IMPORTANCE_DEFAULT);
+        channel.enableLights(false);
+        channel.enableVibration(false);
+        channel.setShowBadge(false);
+
+        notificationManager.createNotificationChannel(channel);
+    }
+
+    static Notification buildOngoingConferenceNotification(boolean isMuted) {
+        Context context = reactInstanceManager.getCurrentReactContext();
```
Comment (Calinteodor) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:80]: Replaced ReactInstanceManagerHolder.

**Comment by saghul** [comment]
How is the service started? On JM we do it on the native side. I don't see hoe it would start here, is there something else missing?

**Comment by saghul** [comment]
> I think we can launch it from react/features/mobile/react-native-sdk/middleware when CONFERENCE_JOINED is triggered.

Yep, but that will need some PR in JM, right?

File: react/features/mobile/react-native-sdk/middleware.js [hunk 2]
```diff
@@ -32,6 +35,7 @@ const externalAPIEnabled = isExternalAPIAvailable();
         rnSdkHandlers?.onConferenceFocused && rnSdkHandlers?.onConferenceFocused();
         break;
     case CONFERENCE_JOINED:
+        JitsiMeetOngoingConferenceService.launch();
```
Comment (Calinteodor) [inline react/features/mobile/react-native-sdk/middleware.js:38]: I used NativeModules to launch it from here. 
Launch method has context and extradata as params, but I think these two can remain undefined, in this case, right?

**Comment by michellavezzo** [comment]
@Calinteodor do you have any idea when this fix will be ready?

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            .setContentText("ongoing_notification_text")
+            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
+            .setContentIntent(pendingIntent)
+            .setOngoing(true)
+            .setWhen(startingTime)
+            .setUsesChronometer(true)
+            .setAutoCancel(false)
+            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
+            .setOnlyAlertOnce(true)
+            .setSmallIcon(context.getResources().getIdentifier("ic_notification", "drawable", context.getPackageName()));
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:100]: Add this icon to the sample project or it won't work.

File: react-native-sdk/android/build.gradle [hunk 4]
```diff
@@ -128,6 +128,7 @@ dependencies {
   implementation 'com.squareup.duktape:duktape-android:1.3.0'
   implementation 'com.dropbox.core:dropbox-core-sdk:4.0.1'
   implementation 'com.jakewharton.timber:timber:4.7.1'
+  implementation 'androidx.localbroadcastmanager:localbroadcastmanager:1.1.0'
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react-native-sdk/android/build.gradle]: Removed.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        super.onDestroy();
+    }
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return null;
+    }
+
+    @Override
+    public void onCurrentConferenceChanged(String conferenceUrl) {
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java]: Removed.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+class JMOngoingConferenceModule
+    extends ReactContextBaseJavaModule {
+
+    public static final String NAME = "JMOngoingConference";
+
+    public JMOngoingConferenceModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    public void createNotification() {
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java]: Exposed.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java (hunk 3 above)
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:100]: Added res folder to prepare_sdk script.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
+
+        NotificationChannel channel
+            = notificationManager.getNotificationChannel("JitsiOngoingConferenceChannel");
+        if (channel != null) {
+            // The channel was already created, no need to do it again.
+            return;
+        }
+
+        channel = new NotificationChannel("JitsiOngoingConferenceChannel", "ongoing_notification_action_unmute", NotificationManager.IMPORTANCE_DEFAULT);
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:65]: @saghul the original OngoingNotification presents it context.getString(R.string.ongoing_notification_action_unmute). If i am not mistaken, regarding mute actions, I can remove this.

File: react/features/mobile/react-native-sdk/middleware.js [hunk 8]
```diff
@@ -32,12 +36,15 @@ const externalAPIEnabled = isExternalAPIAvailable();
...
     case CONFERENCE_JOINED:
+        JMOngoingConference?.launchNotification();
         rnSdkHandlers?.onConferenceJoined && rnSdkHandlers?.onConferenceJoined();
         break;
     case CONFERENCE_LEFT:
+        JMOngoingConference?.destroyNotification();
         //  Props are torn down at this point, perhaps need to leave this one out
         break;
     case CONFERENCE_WILL_JOIN:
+        JMOngoingConference?.createNotification();
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/mobile/react-native-sdk/middleware.js]: Updated, I applied the check for native modules as well for the state listener.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    public static final String NAME = "JMOngoingConference";
+
+    public JMOngoingConferenceModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    @ReactMethod
+    public void launchNotification() {
+        Context context = getReactApplicationContext();
+        Intent screenShareIntent = new Intent(BroadcastEvent.Type.SCREEN_SHARE_TOGGLED.getAction());
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:27]: This is not necessary.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    private static final String EXTRA_DATA_KEY = "extraDataKey";
+    private static final String EXTRA_DATA_BUNDLE_KEY = "extraDataBundleKey";
+
+
+    public static void launch(Context context, HashMap<String, Object> extraData) {
+        RNOngoingNotification.createOngoingConferenceNotificationChannel(context);
+
+        Intent intent = new Intent(context, JitsiMeetOngoingConferenceService.class);
+
+        Bundle extraDataBundle = new Bundle();
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:52]: Drop this.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    public static void abort(Context context) {
+        Intent intent = new Intent(context, JitsiMeetOngoingConferenceService.class);
+        context.stopService(intent);
+    }
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return null;
+    }
+}
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:85]: Here we are missing a `onStartCommand` that actually displays the notification.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        NotificationChannel channel
+            = notificationManager.getNotificationChannel("JitsiOngoingConferenceChannel");
+        if (channel != null) {
+            // The channel was already created, no need to do it again.
+            return;
+        }
+
+        notificationManager.createNotificationChannel(channel);
+    }
+}
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:66]: You need to have some `buildOngoingConferenceNotification` here like we do on native.

File: react/features/mobile/react-native-sdk/middleware.js [hunk 13]
```diff
@@ -56,5 +62,20 @@ const externalAPIEnabled = isExternalAPIAvailable();
...
-}
+});
+
+/**
+ * Check if native modules are being used or not.
+ */
+!externalAPIEnabled && StateListenerRegistry.register(
+    state => state['features/base/conference'].conference,
+    (conference, previousConference) => {
+        if (conference !== previousConference) {
```
Comment (saghul) [inline react/features/mobile/react-native-sdk/middleware.js:73]: Re-write these so we abort if there is no new conference, and launch if there is a new one but there isn't a previous one.

If we jump from one conference to another we need to abort then launch.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    public static final String NAME = "JMOngoingConference";
+
+    private static JMOngoingConferenceModule jmOngoingConferenceModuleInstance;
+
+    public JMOngoingConferenceModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+        jmOngoingConferenceModuleInstance = this;
+    }
+
+    static JMOngoingConferenceModule getInstance() {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:27]: This shouldn't be a singleton, why did you make it so?

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        return jmOngoingConferenceModuleInstance;
+    }
+
+    Notification build() {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:31]: We should be able to pass the right parameters to `launch` so this is not necessary. Why did you need to add this here? Feels kinda out of place since the notification is controlled by the service, not this.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        Context context = getReactApplicationContext().getCurrentActivity();
+
+        return RNOngoingNotification.buildOngoingConferenceNotification(context);
+    }
+
+    void create() {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:37]: Ditto.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+            JitsiMeetLogger.w(TAG + " Couldn't start service, notification is null");
+        } else {
+            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+            notificationManager.notify(RNOngoingNotification.NOTIFICATION_ID, notification);
+        }
+
+        return START_NOT_STICKY;
+    }
+
+    public static void onCurrentConferenceChanged(String conferenceUrl, Context context) {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:111]: This is unneccesary, abort should be called.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+    @ReactMethod
+    public void abort() {
+        Context context = getReactApplicationContext();
+
+        JitsiMeetOngoingConferenceService.abort(context);
+    }
+
+    @ReactMethod
+    public void onCurrentConferenceChanged() {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:58]: This is unnecessary, see below.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+ * The service will help keep the app running while in the background.
+ *
+ * See: https://developer.android.com/guide/components/services
+ */
+public class JitsiMeetOngoingConferenceService extends Service {
+    private static final String TAG = JitsiMeetOngoingConferenceService.class.getSimpleName();
+
+    public static void launch(Context context) {
+
+        JMOngoingConferenceModule.getInstance().create();
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:46]: You can extract the activity from the context here and pass it to the function `create` calls. That indirection is messy and not necessary.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    public static void abort(Context context) {
+        Intent intent = new Intent(context, JitsiMeetOngoingConferenceService.class);
+        context.stopService(intent);
+    }
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+
+        Notification notification = JMOngoingConferenceModule.getInstance().build();
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:79]: Are you sure we cannot obtain what we need to build the notification?

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Helper class for creating the ongoing notification which is used with
+ * {@link JitsiMeetOngoingConferenceService}. It allows the user to easily get back to the app
+ * and to hangup from within the notification itself.
+ */
+class RNOngoingNotification {
+    private static final String TAG = RNOngoingNotification.class.getSimpleName();
+
+    static final int NOTIFICATION_ID = new Random().nextInt(99999) + 10000;
+    private static long startingTime = 0;
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:43]: Drop this, it's not needed.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java (hunk 20 above)
Comment (saghul) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:79]: I think you should be able to get the current context here, in order to pass it around.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return null;
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+
+        Notification notification = JMOngoingConferenceModule.getInstance().build();
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:98]: I don't think you need to do anything here, since the notification was created before. Just return `START_NOT_STICKY`.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+            JitsiMeetLogger.w(TAG + " Cannot create notification: no current context");
+            return null;
+        }
+
+        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, "JitsiOngoingConferenceChannel");
+
+        Intent notificationIntent = new Intent(context, context.getClass());
+        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, notificationIntent, PendingIntent.FLAG_IMMUTABLE);
+
+        if (startingTime == 0) {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:84]: Drop this and just set the value directly in the call to `setWhen`

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+            .setUsesChronometer(true)
+            .setAutoCancel(false)
+            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
+            .setOnlyAlertOnce(true)
+            .setSmallIcon(context.getResources().getIdentifier("ic_notification", "drawable", context.getPackageName()));
+
+        return builder.build();
+    }
+
+    static void resetStartingtime() {
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/RNOngoingNotification.java:105]: Drop this method.

File: react/features/mobile/react-native-sdk/middleware.js [hunk 25]
```diff
@@ -56,5 +62,21 @@ const externalAPIEnabled = isExternalAPIAvailable();
...
+ */
+!externalAPIEnabled && StateListenerRegistry.register(
+    state => state['features/base/conference'].conference,
+    (conference, previousConference) => {
+        if (!conference) {
+            JMOngoingConference.abort();
+        } else if (conference && !previousConference) {
+            JMOngoingConference.launch();
+        } else if (conference !== previousConference) {
+            JMOngoingConference.onCurrentConferenceChanged();
```
Comment (saghul) [inline react/features/mobile/react-native-sdk/middleware.js:78]: Just abort should be enough.

**Review by saghul — CHANGES_REQUESTED** [review]
Almost there!

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java (hunk 20 above)
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:79]: build() is part of methods that override the Service methods and those aren't static methods by default. I think it's easier to tackle this situation with an instance.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java (hunk 22 above)
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:98]: I'll check to see if that is correct.

Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:98]: We need to do it here as well. Build fails if we don't.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java (hunk 20 above)
Comment (saghul) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:79]: I'd like to explore the other option pl.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java (hunk 16 above)
Comment (Calinteodor) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:37]: So we need to keep this method here because this is a context module from which we need to get the activity.
If we do it in another place  error: cannot find symbol
        Context activityContext = context.getCurrentActivity();

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java (hunk 22 above)
Comment (saghul) [reply react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:98]: It doesn't maske sense. What build? We are essentially posting exactly the same notification twixce. We shouldn't need to.

File: react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+    public static final String NAME = "JMOngoingConference";
+
+    public JMOngoingConferenceModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    @ReactMethod
+    public void launch() {
+        Context context = getReactApplicationContext();
+        Context activityContext = getReactApplicationContext().getCurrentActivity();
```
Comment (saghul) [inline react-native-sdk/android/src/main/java/org/jitsi/meet/sdk/JMOngoingConferenceModule.java:27]:
```suggestion
        Context activityContext = context.getCurrentActivity();
```

**Review by saghul — APPROVED** [review]
tint nit, LGTM

**Review by saghul — APPROVED** [review]
